pkg/sql/authorization.go

```diff
@@ -940,3 +943,104 @@ func insufficientPrivilegeError(
 		"user %s does not have %s privilege on %s %s",
 		user, kind, typeForError, object.GetName())
 }
+
+// WarmSyntheticPrivilegeCacheForVirtualTables will attempt to access
+// each virtual table, and, in doing so, will cache the privileges for those
+// virtual tables. This work is done in parallel. It aids in reducing the cold
+// start latency when accessing virtual tables which themselves list virtual
+// tables.
+func WarmSyntheticPrivilegeCacheForVirtualTables(
+	ctx context.Context,
+	stopper *stop.Stopper,
+	version clusterversion.Handle,
+	cache *cacheutil.Cache,
+	schemas catalog.VirtualSchemas,
+	ief descs.TxnManager,
+	db *kv.DB,
+) {
+	if !version.IsActive(ctx, clusterversion.V22_2SystemPrivilegesTable) {
+		return
+	}
+
+	start := timeutil.Now()
+	var err error
+	defer func() {
+		if err != nil {
+			log.Warningf(ctx, "failed to warm privileges for virtual tables: %v", err)
+		} else {
+			log.Infof(ctx, "warmed privileges for virtual tables in %v", timeutil.Since(start))
+		}
+	}()
+
+	var tableVersions []descpb.DescriptorVersion
+	vtablePathToPrivilegeDescs := make(map[string]*catpb.PrivilegeDescriptor)
+	query := fmt.Sprintf(
+		`SELECT path, username, privileges, grant_options FROM system.%s WHERE path LIKE '/%s/%%'`,
+		catconstants.SystemPrivilegeTableName,
+		syntheticprivilege.VirtualTablePathPrefix,
+	)
+	err = ief.DescsTxnWithExecutor(ctx, db, nil /* sessionData */, func(ctx context.Context, txn *kv.Txn, descsCol *descs.Collection, ie sqlutil.InternalExecutor) (retErr error) {
+		_, systemPrivDesc, err := descsCol.GetImmutableTableByName(
+			ctx,
+			txn,
+			syntheticprivilege.SystemPrivilegesTableName,
+			tree.ObjectLookupFlagsWithRequired(),
+		)
+		if err != nil {
+			return err
+		}
+		if systemPrivDesc.IsUncommittedVersion() {
+			// This shouldn't ever happen, but if it does somehow, then we can't pre-warm the cache.
+			return errors.Newf("%s is at an uncommitted version", syntheticprivilege.SystemPrivilegesTableName)
+		}
+		tableVersions = []descpb.DescriptorVersion{systemPrivDesc.GetVersion()}
+
+		it, err := ie.QueryIteratorEx(
+			ctx, `get-vtable-privileges`, txn, sessiondata.NodeUserSessionDataOverride, query,
+		)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			retErr = errors.CombineErrors(retErr, it.Close())
+		}()
+
+		for {
+			ok, err := it.Next(ctx)
+			if err != nil {
+				return err
+			}
+			if !ok {
+				break
+			}
+			path := tree.MustBeDString(it.Cur()[0])
+			user := tree.MustBeDString(it.Cur()[1])
+			privArr := tree.MustBeDArray(it.Cur()[2])
+			grantOptionArr := tree.MustBeDArray(it.Cur()[3])
+			privileges, err := synthesizePrivilegeDescriptorFromSystemPrivilegesRow(privilege.VirtualTable, user, privArr, grantOptionArr)
```

Comment:
It seems like there's a row in the `system.privileges` table per user, but then this overwrites the privilege descriptor for each row. Am I missing something, or do we need to collect the rows for all of the users for a given table together?

```
grant select on crdb_internal.tables to root;
grant select on crdb_internal.tables to foo; 
select * from system.privileges;                                                             
```
```
  username |             path             | privileges | grant_options
-----------+------------------------------+------------+----------------
  foo      | /vtable/crdb_internal/tables | {SELECT}   | {}
  root     | /vtable/crdb_internal/tables | {SELECT}   | {}

Reply:
oh i see. i missed that in the existing usage of `synthesizePrivilegeDescriptorFromSystemPrivileges` -- it makes a `*PrivilegeDescriptor` and accumulates everything into that.